Describe why check_min_cppstd is not enough (#5146)
conan-io/conan-center-index: docs/faqs.md

  * [Why PDB files are not allowed?](#why-pdb-files-are-not-allowed)
      * [Why is there no option for PDB, as there is for fPIC?](#why-is-there-no-option-for-pdb-as-there-is-for-fpic)
  * [Why _installer_ packages remove some settings from their package ID?](#why-_installer_-packages-remove-some-settings-from-their-package-id)
  * [Can I remove an option from recipe](#can-i-remove-an-option-from-recipe)

>   * [Why is `tools.check_min_cppstd call` not enough?](#why-is-toolscheck_min_cppstd-call-not-enough)<!-- endToc -->

Reviewer: ```suggestion
  * [Why is a `tools.check_min_cppstd` call not enough?](#why-is-toolscheck_min_cppstd-call-not-enough)<!-- endToc -->
```

      * [But if there are no packages available, what will the x86 validation look like?](#but-if-there-are-no-packages-available-what-will-the-x86-validation-look-like)
  * [Why PDB files are not allowed?](#why-pdb-files-are-not-allowed)
      * [Why is there no option for PDB, as there is for fPIC?](#why-is-there-no-option-for-pdb-as-there-is-for-fpic)
  * [Why _installer_ packages remove some settings from their package ID?](#why-installer-packages-remove-some-settings-from-their-package-id)
  * [Can I remove an option from recipe](#can-i-remove-an-option-from-recipe)
  * [What license should I use for Public Domain?](#what-license-should-i-use-for-public-domain)

>   * [Why is a `tools.check_min_cppstd` call not enough?](#why-is-toolscheck_min_cppstd-call-not-enough)<!-- endToc -->

Reviewer 2: ```suggestion
  * [Why is a `tools.check_min_cppstd` call not enough?](#why-is-toolscheck_min_cppstd-call-not-enough)
<!-- endToc -->
```

I think this should work, but looks better.

Author: The tooling auto does this any ways

However, if logic is too complex (this is subjective and depends on the Conan review team) then just remove the option.
After one month, we will welcome a PR removing the option that was deprecated.

> ## Why is `tools.check_min_cppstd call` not enough?

Reviewer: ```suggestion
## Why is a `tools.check_min_cppstd` call not enough?
```

> Very often C++ projects require a minimum standard version, such as 14 or 17, in order to compiler. Conan offers tools which enable checking this setting is enabled above this a certain version otherwise it uses the compiler's default.

Reviewer: I feel that these changes would improve overall clarity:
```suggestion
Very often C++ projects require a minimum standard version, such as 14 or 17, in order to compile. Conan offers tools which enable checking the relevant setting is enabled and above this support for a certain version is present. Otherwise, it uses the compiler's default.
```

```python
def configure(self):
    tools.check_min_cppstd(self, 14) 👈 Wrong!
```

This fails to cover the waste number of use cases for the following reasons:

1. `cppstd` is not configured in the `--detect`ed profiles generated by Conan, the majority of users simply do not have this setting.
2. A shocking number of projects override this setting within their respective build scripts, this setting does not get applied in those cases

> 3. Conan-Center-Index too does manage the `cppstd` setting for the compilers it supports to generate binaries.

Author: ```suggestion
3. Conan-Center-Index does **not** manage the `cppstd` setting for the compilers it supports to generate binaries.
```

## Why is a `tools.check_min_cppstd` call not enough?

Very often C++ projects require a minimum standard version, such as 14 or 17, in order to compile. Conan offers tools which enable checking the relevant setting is enabled and above this support for a certain version is present. Otherwise, it uses the compiler's default.

```python
def configure(self):
    tools.check_min_cppstd(self, 14) 👈 Wrong!
```

This fails to cover the waste number of use cases for the following reasons:

1. `cppstd` is not configured in the `--detect`ed profiles generated by Conan, the majority of users simply do not have this setting.
2. A shocking number of projects override this setting within their respective build scripts, this setting does not get applied in those cases
3. Conan-Center-Index does **not** manage the `cppstd` setting for the compilers it supports to generate binaries.

```python

> def configure(self):

Reviewer 3: ```suggestion
def validate(self):
```

However, if logic is too complex (this is subjective and depends on the Conan review team) then just remove the option.
After one month, we will welcome a PR removing the option that was deprecated.

> 

Reviewer 2: Please, add this line back! :)

[The Public Domain](https://fairuse.stanford.edu/overview/public-domain/welcome/) is not a license by itselt. Thus, we have [equivalent licenses](https://en.wikipedia.org/wiki/Public-domain-equivalent_license) to be used instead. By default, if a project uses Public Domain and there is no offcial license listed, you should use [Unlicense](https://spdx.org/licenses/Unlicense).

## Why is a `tools.check_min_cppstd` call not enough?

Very often C++ projects require a minimum standard version, such as 14 or 17, in order to compile. Conan offers tools which enable checking the relevant setting is enabled and above this support for a certain version is present. Otherwise, it uses the compiler's default.

```python
def configure(self):
    tools.check_min_cppstd(self, 14) 👈 Wrong!
```

This fails to cover the waste number of use cases for the following reasons:

1. `cppstd` is not configured in the `--detect`ed profiles generated by Conan, the majority of users simply do not have this setting.

> 2. A shocking number of projects override this setting within their respective build scripts, this setting does not get applied in those cases

Reviewer 2: ```suggestion
2. A shocking number of projects override this setting within their respective build scripts, this setting does not get applied in those cases.
```

Reviewer 2: Does the correct solution solve this issue?
